File: typescript/src/graph/dinic.ts
// LombokAlgoritma — Maximum flow (Dinic)
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok
import { InvalidInputError, NegativeWeightError } from '../core/errors.js';
import { type Graph, assertNode, validateGraph } from './types.js';

/**
 * Value of a maximum `source → sink` flow; edge weights are capacities (parallel edges add up).
 * Dinic's algorithm: BFS level graph + blocking flow by DFS with current-arc pointers.
 * O(V²·E). Only the flow *value* is normative (SPEC §6.13); it is exact for integer capacities
 * below 2^53.
 *
 * @throws RangeError on a negative capacity or source = sink
 */
export function dinic(g: Graph, source: number, sink: number): number {
  validateGraph(g);
  assertNode(g, source, 'source');
  assertNode(g, sink, 'sink');
  if (source === sink) throw new InvalidInputError('dinic: source and sink must differ');
  const n = g.nodes;
  // residual graph as parallel arrays; edge i and i ^ 1 are a forward/backward pair
  const to: number[] = [];
  const cap: number[] = [];
  const head: number[][] = Array.from({ length: n }, () => []);
  for (const e of g.edges) {
    if (e.weight < 0) throw new NegativeWeightError('dinic');
    (head[e.from] as number[]).push(to.length);
    to.push(e.to);
    cap.push(e.weight);
    (head[e.to] as number[]).push(to.length);
    to.push(e.from);
    cap.push(0);
  }
  const level = new Array<number>(n);
  const it = new Array<number>(n);
  const bfs = (): boolean => {
    level.fill(-1);
    level[source] = 0;
    const q = [source];
    for (let h = 0; h < q.length; h++) {
      const u = q[h] as number;
      for (const id of head[u] ?? []) {
        const v = to[id] as number;
        if ((cap[id] as number) > 0 && level[v] === -1) {
          level[v] = (level[u] as number) + 1;
          q.push(v);
        }
      }
    }
    return level[sink] !== -1;
  };
  // iterative blocking-flow DFS (augment one path at a time)
  const augment = (): number => {
    const path: number[] = []; // edge ids
    let u = source;
    for (;;) {
      if (u === sink) {
        let f = Number.POSITIVE_INFINITY;
        for (const id of path) f = Math.min(f, cap[id] as number);
        for (const id of path) {
          cap[id] = (cap[id] as number) - f;
          cap[id ^ 1] = (cap[id ^ 1] as number) + f;
        }
        return f;
      }
      const edges = head[u] as number[];
      let advanced = false;
      while ((it[u] as number) < edges.length) {
        const id = edges[it[u] as number] as number;
        const v = to[id] as number;
        if ((cap[id] as number) > 0 && level[v] === (level[u] as number) + 1) {
          path.push(id);
          u = v;
          advanced = true;
          break;
        }
        it[u] = (it[u] as number) + 1;
      }
      if (advanced) continue;
      if (u === source) return 0;
      level[u] = -1; // dead end: prune
      const back = path.pop() as number;
      u = to[back ^ 1] as number;
      it[u] = (it[u] as number) + 1;
    }
  };
  let flow = 0;
  while (bfs()) {
    it.fill(0);
    for (let f = augment(); f > 0; f = augment()) flow += f;
  }
  return flow;
}

File: rust/lombokalgoritma/src/graph/dinic.rs
// LombokAlgoritma — Dinic maximum flow (SPEC §9.11)
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok
use super::types::{check_node, validate, Graph};
use crate::{Error, Result};
use alloc::vec;
use alloc::vec::Vec;

struct Residual {
    to: Vec<usize>,
    cap: Vec<f64>,
    head: Vec<Vec<usize>>,
    level: Vec<Option<usize>>,
    it: Vec<usize>,
}

impl Residual {
    fn bfs(&mut self, s: usize, t: usize) -> bool {
        self.level.fill(None);
        self.level[s] = Some(0);
        let mut q = vec![s];
        let mut h = 0;
        while h < q.len() {
            let u = q[h];
            h += 1;
            let lu = self.level[u].unwrap_or(0);
            for &id in &self.head[u] {
                let v = self.to[id];
                if self.cap[id] > 0.0 && self.level[v].is_none() {
                    self.level[v] = Some(lu + 1);
                    q.push(v);
                }
            }
        }
        self.level[t].is_some()
    }

    /// Find one augmenting path in the level graph and push flow along it; 0 when blocked.
    fn augment(&mut self, s: usize, t: usize) -> f64 {
        let mut path: Vec<usize> = Vec::new();
        let mut u = s;
        loop {
            if u == t {
                let f = path
                    .iter()
                    .fold(f64::INFINITY, |f, &id| f.min(self.cap[id]));
                for &id in &path {
                    self.cap[id] -= f;
                    self.cap[id ^ 1] += f;
                }
                return f;
            }
            let mut advanced = false;
            while self.it[u] < self.head[u].len() {
                let id = self.head[u][self.it[u]];
                let v = self.to[id];
                let next_level = self.level[u].map(|l| l + 1);
                if self.cap[id] > 0.0 && self.level[v].is_some() && self.level[v] == next_level {
                    path.push(id);
                    u = v;
                    advanced = true;
                    break;
                }
                self.it[u] += 1;
            }
            if advanced {
                continue;
            }
            if u == s {
                return 0.0;
            }
            self.level[u] = None; // dead end: prune
            let Some(back) = path.pop() else {
                return 0.0;
            };
            u = self.to[back ^ 1];
            self.it[u] += 1;
        }
    }
}

/// Value of a maximum `source → sink` flow; edge weights are capacities (parallel edges add up).
/// BFS level graph + blocking flow with current-arc pointers, O(V²·E). Only the flow value is
/// normative; it is exact for integer capacities below 2^53.
///
/// # Errors
/// [`Error::OutOfRange`] for an edge endpoint, `source` or `sink` outside `[0, n)`;
/// [`Error::InvalidInput`] when `source == sink`; [`Error::NegativeWeight`] for a negative
/// capacity.
pub fn dinic(g: &Graph, source: usize, sink: usize) -> Result<f64> {
    validate(g)?;
    check_node(g, source)?;
    check_node(g, sink)?;
    if source == sink {
        return Err(Error::InvalidInput);
    }
    let n = g.nodes;
    let mut r = Residual {
        to: Vec::with_capacity(2 * g.edges.len()),
        cap: Vec::with_capacity(2 * g.edges.len()),
        head: vec![Vec::new(); n],
        level: vec![None; n],
        it: vec![0; n],
    };
    for e in &g.edges {
        if e.weight < 0.0 {
            return Err(Error::NegativeWeight);
        }
        r.head[e.from].push(r.to.len());
        r.to.push(e.to);
        r.cap.push(e.weight);
        r.head[e.to].push(r.to.len());
        r.to.push(e.from);
        r.cap.push(0.0);
    }
    let mut flow = 0.0;
    while r.bfs(source, sink) {
        r.it.fill(0);
        loop {
            let f = r.augment(source, sink);
            if f > 0.0 {
                flow += f;
            } else {
                break;
            }
        }
    }
    Ok(flow)
}
